cmd/selfupdatectl/upload.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"log"
+	"runtime"
+
+	"github.com/fynelabs/selfupdate/cmd/selfupdatectl/internal/cloud"
+	"github.com/urfave/cli/v2"
+)
+
+func upload() *cli.Command {
+	var endpoint string
+	var region string
+	var bucket string
+	var akid string
+	var secret string
+	var baseS3Path string
+	var binaryPattern string
+
+	a := &application{}
+
+	return &cli.Command{
+		Name:        "upload",
+		Usage:       "Upload an executable which has been properly signed to S3",
+		Description: "The executable specified will get its signature generated and checked before being uploaded to a S3 bucket than can be optionally specified.",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:        "private-key",
+				Aliases:     []string{"priv"},
+				Usage:       "The private key file to store the new key in.",
+				Destination: &a.privateKey,
+				Value:       "ed25519.key",
+			},
+			&cli.StringFlag{
+				Name:        "public-key",
+				Aliases:     []string{"pub"},
+				Usage:       "The public key file to store the new key in.",
+				Destination: &a.publicKey,
+				Value:       "ed25519.pem",
+			},
+			&cli.StringFlag{
+				Name:        "aws-endpoint",
+				Aliases:     []string{"e"},
+				Usage:       "AWS endpoint to connect to (can be used to connect to non AWS S3 services)",
+				EnvVars:     []string{"AWS_S3_ENDPOINT"},
+				Destination: &endpoint,
+			},
+			&cli.StringFlag{
+				Name:        "aws-region",
+				Aliases:     []string{"r"},
+				Usage:       "AWS region to connect to",
+				EnvVars:     []string{"AWS_S3_REGION"},
+				Destination: &region,
+			},
+			&cli.StringFlag{
+				Name:        "aws-bucket",
+				Aliases:     []string{"b"},
+				Usage:       "AWS bucket to store data into",
+				EnvVars:     []string{"AWS_S3_BUCKET"},
+				Destination: &bucket,
+			},
+			&cli.StringFlag{
+				Name:        "aws-secret",
+				Aliases:     []string{"s"},
+				Usage:       "AWS secret to use to establish S3 connection",
+				Destination: &secret,
+			},
+			&cli.StringFlag{
+				Name:        "aws-AKID",
+				Aliases:     []string{"a"},
+				Usage:       "AWS Access Key ID to use to establish S3 connection",
+				Destination: &akid,
+			},
+			&cli.StringFlag{
+				Name:        "aws-base-s3-path",
+				Aliases:     []string{"path", "p"},
+				Usage:       "Specify the sub path in which the binary will be uploaded",
+				Destination: &baseS3Path,
+			},
+			&cli.StringFlag{
+				Name:        "binary-template",
```

Comment:
This name confuses me a little, as everywhere else it is called executable, not binary.
Maybe "template" is enough.
One thing I am not sure about is if the subcommand "upload" is S3 specific, if so if we may add other upload targets maybe it should be "s3upload".
If this is indeed aws s3 specific then all of the parameters having "aws" prefix seems like duplication (and maybe the s3 mention in base path too).

Reply:
It is s3 specific at the moment, but I am hoping that we can have parameters to set different upload mode and have one name. I can easily rename to s3upload for the moment and when we add new mode we can merge them then.